Optimization: Add def use analysis

antiagainst opened the pull request.

Qining commented:
Hope to have `GetDef()` and `GetUses()` so that user won't access the map directly.

Qining commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,571 @@
+              {8, {"OpBranchConditional %8 %9 %7"}},
+              {9, {"OpBranchConditional %8 %9 %7"}},
+            }
+          }
```
What about add some tests for forward-reference?

antiagainst replied:
Done.

Qining commented on source/opt/def_use_manager.h:
```diff
@@ -0,0 +1,81 @@
+  // Turns the instruction defining the given |id| into a Nop. Returns true on
+  // success, false if the given |id| is not defined at all. This method does
+  // not erase the uses of |id|.
+  bool KillDef(uint32_t id);
```
I think we may have a check of the uses for the deleting id here? This is easy to check and looks useful to me.

antiagainst replied:
You mean return some error code to show that the id to be deleted still have uses?

Qining replied:
Oops, never mind, now I think that is not necessary.

antiagainst commented:
Makes sense, I'll add `GetUses` and `GetDef`.

antiagainst commented:
Added `GetDef` and `GetUses`.

Qining commented:
LGTM +1

dneto0 commented on source/opt/def_use_manager.cpp:
```diff
@@ -0,0 +1,93 @@
+#include <cassert>
+#include <functional>
+
+#include "def_use_manager.h"
```
This should go first.  It's the header that defines the interface for this module.

antiagainst replied:
Done.

dneto0 commented on source/opt/def_use_manager.h:
```diff
@@ -0,0 +1,89 @@
+namespace spvtools {
+
+namespace ir {
+class Instruction;
```
FYI.  Google style guide discourages forward declarations.
Ok to keep for now.

antiagainst replied:
Done.

dneto0 commented on source/opt/def_use_manager.h:
```diff
@@ -0,0 +1,89 @@
+
+  // Analyzes the defs and uses in the given |module| and populates data
+  // structures in this class.
+  void AnalyzeDefUse(ir::Module* module);
```
Since this is just an analysis, I expect that Module will not be changed.  So I expected to see the argument as  const ir::Module& module

But you can't do that since there is no const overload of Module::ForEachInst(const std::function<void(const Instruction*)> & ?

antiagainst replied:
Right... I'll add the const overload later.

dneto0 commented on source/opt/def_use_manager.h:
```diff
@@ -0,0 +1,89 @@
+
+  // Turns the instruction defining the given |id| into a Nop. Returns true on
+  // success, false if the given |id| is not defined at all. This method does
+  // not erase the uses of |id|.
```
Does this method also remove all record of uses of this definition?

antiagainst replied:
Now it does.

dneto0 commented on source/opt/def_use_manager.h:
```diff
@@ -0,0 +1,89 @@
+// A class for analyzing and managing defs and uses in an ir::Module.
+class DefUseManager {
+ public:
+  using InstLocPair = std::pair<ir::Instruction*, uint32_t>;
```
Should this be called "Use" instead?  It feels like we should have a more first-class concept of a use.

dneto0 replied:
Actually, it's really unclear what the uint32_t is.  Is it the logical operand Id?  Could it be the type Id for the target instruction?

antiagainst replied:
Added a `Use` struct and comments for it.

dneto0 commented on source/opt/def_use_manager.h:
```diff
@@ -0,0 +1,89 @@
+ public:
+  using InstLocPair = std::pair<ir::Instruction*, uint32_t>;
+  using IdToDefMap = std::unordered_map<uint32_t, ir::Instruction*>;
+  using IdToUseMap = std::unordered_map<uint32_t, std::list<InstLocPair>>;
```
I strongly suggest making a name for the set-of-uses type.  Especially since std::list will be very slow when there are a lot of elements.

dneto0 replied:
IdToUseMap -> IdToUsesMap

dneto0 replied:
Speaking of which, why use a std::list?  There should be a very good reason to do so, e..g if you want iterators relatively stable.

antiagainst replied:
Done.

`list` is good for removing in the middle. Needed in the new impl for `KillDef`.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+
+  std::string text;
+  // We'll need to check the underlying id numbers.
+  // So turn off friendly names for ids.
```
:-)

Qining replied:
I think it would be clearer with && ~SPV_BINARY_TO_TEXT_OPTION_FRIENDLY_NAMES

antiagainst replied:
Added.

dneto0 commented on source/opt/def_use_manager.cpp:
```diff
@@ -0,0 +1,93 @@
+    // TODO(antiagainst): de-duplication.
+    id_to_uses_[after].emplace_back(inst, loc);
+  }
+  id_to_uses_.erase(before);
```
This is wrong if before == after.
If before==after then this should be a no-op.  (That might be worth mentioning in the blurb for the function.)

dneto0 replied:
And we'll need a test case for this.

antiagainst replied:
Good catch! Done. Test case added.

dneto0 commented on source/opt/def_use_manager.cpp:
```diff
@@ -0,0 +1,93 @@
+
+  for (uint32_t i = 0; i < inst->NumOperands(); ++i) {
+    const auto type = inst->GetOperand(i).type;
+    if (type == SPV_OPERAND_TYPE_ID || type == SPV_OPERAND_TYPE_TYPE_ID ||
```
Worth having a comment here: any Id operand that is not a result-id.
(Also, I prefer switches in this kind of code...)

antiagainst replied:
Done. Changed to `switch`.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+        "OpMemoryModel Logical GLSL450",
+        {},
+      },
+      { // replace one use
```
Add test case:  Replace an Id with itself.

antiagainst replied:
Done.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+            {20, {"%5 = OpTypeMatrix %20 2"}},
+            {3, {"%6 = OpTypeMatrix %3 3"}},
+            {40, {"%7 = OpTypeMatrix %40 4"}},
+          },
```
Another test case:
  Replace a case Id in OpSwitch over 64-bit integers.  (other than the default case)
This is interesting because it will check that you are replacing the right word when the literals are 2 words wide.

(We also don't have lower level tests for SetOperand and GetSingleWordOperand.  But at least a case here will check it at slightly above unit test level.)

antiagainst replied:
I've a spotcheck test for it now.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+            {1,
+              {
+                "%2 = OpTypeVector %1 2",
+                "OpNop", // %3 = OpTypeVector %1 3
```
This is unexpected.  The OpNop is no longer a use of 1. 
KillDef should remove the uses , I think.

antiagainst replied:
No longer like this.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+            },
+            {3,
+              {
+                "OpNop" // %5 = OpTypeMatrix %3 3
```
Also will need a test case where killing a def will make someone's uses list empty, in which case that list should be removed?

antiagainst replied:
Now the `OpPhi` test case demonstrates this.

dneto0 commented:
Review complete.

Qining commented on source/opt/def_use_manager.cpp:
```diff
@@ -0,0 +1,93 @@
+  return id_to_defs_.at(id);
+}
+
+std::list<DefUseManager::InstLocPair>* DefUseManager::GetUses(uint32_t id) {
```
Maybe not necessary in this CL. But it would be better if we define a type for std::list<DefUseManager::InstLocPair>

antiagainst replied:
Done.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+INSTANTIATE_TEST_CASE_P(
+    TestCase, ParseDefUseTest,
+    ::testing::ValuesIn(std::vector<ParseDefUseCase>{
+        {"", {{}, {}}},                              // no instruction
```
Please add a test case where an OpPhi is both the definintion and use of a value.

Note that at some point LLVM decided that mention of a value on the right-hand-side of a phi is *not* considered a use.  But our code does consider it a use.
(And we might want to make that clear in the comments for the DefUseManager)

antiagainst replied:
Done.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+      { // no use, no replace request
+        "", {}, "", {},
+      },
+      { // no use, some replace requests
```
Please add a test case where an OpPhi both produces and consumes a value.

antiagainst replied:
Done.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -0,0 +1,599 @@
+      { // no def, no use, no kill
+        "", {}, "", {}
+      },
+      { // kill nothing
```
Add a test case where an OpPhi both produces and consumes a value.

antiagainst replied:
Done.

dneto0 commented:
(Very nice test cases, by the way!)

antiagainst commented:
Addressed all comments in the fourth commit.

dneto0 commented on source/opt/def_use_manager.h:
```diff
@@ -32,54 +32,70 @@
-  IdToUseMap id_to_uses_;  // Mapping from ids to their uses
+  IdToDefMap id_to_def_;       // Mapping from ids to their definitions
+  IdToUsesMap id_to_uses_;     // Mapping from ids to their uses
+  IdToParamMap id_to_params_;  // Mapping from ids to those ids used as operands
```
"as operands" -> "as function parameters" ?

dneto0 replied:
Hm.  Now that I look at the code to build this, it's a map from the Ids of instructions that use an id, to the list of ids they use, and it can contain duplicates.
Better to call this id_to_used_ids_ or at least have a better comment.

antiagainst replied:
Good call. I've revised the name and comments. It should be more clear now!

dneto0 commented on source/opt/def_use_manager.cpp:
```diff
@@ -42,48 +42,76 @@ void DefUseManager::AnalyzeDefUse(ir::Module* module) {
+        uint32_t use_id = inst->GetSingleWordOperand(i);
+        // use_id is used by the instruction generating def_id.
+        id_to_uses_[use_id].push_back({inst, i});
+        id_to_params_[def_id].push_back(use_id);
```
What if def_id is 0?  This will happen if the current instruction is something like OpDecorate (which does not return a result id).
Seem wrong.

antiagainst replied:
Good catch!! I must be too tired when coding this. Fixed.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -47,7 +52,9 @@ std::string DisassembleInst(ir::Instruction* inst) {
-  tools.Disassemble(binary, &text, SPV_BINARY_TO_TEXT_OPTION_NO_HEADER);
+  tools.Disassemble(binary, &text,
+                    SPV_BINARY_TO_TEXT_OPTION_NO_HEADER &
+                        ~SPV_BINARY_TO_TEXT_OPTION_FRIENDLY_NAMES);
```
I'm surprised you have to explicitly exclude FRIENDLY_NAMES since it's not the default. ?

antiagainst replied:
It's to make it more clear. But I don't care that much. Will remove.

antiagainst replied:
Done.

dneto0 commented on test/opt/test_def_use.cpp:
```diff
@@ -595,5 +884,123 @@ INSTANTIATE_TEST_CASE_P(
 );
 // clang-format on
+//
+TEST(DefUseTest, OpSwith) {
```
"OpSwitch"

antiagainst replied:
:)

dneto0 commented:
I'd like to see a case where the same instruction both defines and uses an Id.  This is valid for an OpPhi.

E.g.
  
```
  %entry = OpLabel
   %zero = OpConstantZero
   OpBranch %loop

 %loop = OpLabel
     %i = OpPhi %i32 %i %loop %zero %entry ; Both defines and uses %i  !!
   OpBranch %loop

antiagainst commented:
Cool code snippet!

Didn't anticipated that you can do this kind of work using `OpPhi`. So `OpPhi` is indeed the puppy for compiler testing. ;)

Added.

Addressed all comments thus far.

dneto0 commented:
+2
Thanks for the TODO.  :-)
